### PR TITLE
test(validator): add coverage for registry and service modules

### LINKED ISSUE #493: test: add coverage for validator registry and service (both at 0%)
## Summary

`src/lean_spec/subspecs/validator/registry.py` and `src/lean_spec/subspecs/validator/service.py` both have **0% test coverage** — no test files exist. These are actively used production modules (imported by `__main__.py`, `node.py`, and each other) that manage validator key loading and drive block/attestation production.

**Coverage stats:**
| File | Statements | Covered | Coverage |
|------|-----------|---------|----------|
| `validator/registry.py` | 118 | 0 | **0%** |
| `validator/service.py` | 159 | 0 | **0%** |

#### `ValidatorManifestEntry` (Pydantic model)
- **Field validation**: Verify `parse_pubkey_hex` converts integer pubkeys to zero-padded hex strings (YAML parsers may auto-convert `0x...` to int)
- **String passthrough**: Verify hex string pubkeys pass through unchanged
- **Model construction**: Verify all fields (index, attestation/proposal pubkey hex, privkey files) are stored correctly

#### `ValidatorManifest` (Pydantic model)
- **YAML loading**: Test `from_yaml_file()` with a valid manifest file (use `tmp_path` fixture)
- **Field validation**: Verify all metadata fields (key_scheme, hash_function, encoding, lifetime, etc.)
- **Validators list**: Verify nested `ValidatorManifestEntry` objects are parsed correctly

#### `load_node_validator_mapping()`
- **Normal loading**: YAML with multiple nodes mapping to validator indices
- **Empty file**: Returns empty dict (YAML parses to `None`)
- **Single node**: One node with multiple indices

#### `ValidatorRegistry` (dataclass)
- **`add()` / `get()`**: Add entries and retrieve by index
- **`get()` miss**: Returns `None` for unknown index
- **`__contains__`**: Membership check for known/unknown indices
- **`indices()`**: Returns `ValidatorIndices` with all registered indices
- **`primary_index()`**: Returns first index, or `None` for empty registry
- **`__len__`**: Correct count after adds
- **`from_yaml()`**: Integration test loading from real YAML files (use `tmp_path`)
  - Happy path: node with assigned validators
  - Unknown node: returns empty registry
  - Missing manifest entry: logs warning, skips validator
  - Missing key file: raises `ValueError`
  - Corrupt key file: raises `ValueError`
- **`from_secret_keys()`**: Convenience constructor with dict of key pairs

#### `ValidatorEntry` (frozen dataclass)
- **Construction**: Verify index and both secret keys are stored
- **Immutability**: Frozen dataclass rejects attribute assignment

#### `ValidatorService` — Properties and lifecycle
- **`stop()` / `is_running`**: Verify running flag transitions
- **`blocks_produced` / `attestations_produced`**: Counter properties start at 0

#### `ValidatorService._sign_block()`
- **Happy path**: Signs block root with proposal key, returns `SignedBlock` with correct structure
- **Unknown validator**: Raises `ValueError`
- **Signature structure**: Verify `BlockSignatures` contains proposer signature and attestation signatures

#### `ValidatorService._sign_attestation()`
- **Happy path**: Signs attestation data root with attestation key, returns `SignedAttestation`
- **Unknown validator**: Raises `ValueError`
- **Correct fields**: `validator_id`, `data`, and `signature` match expectations

#### `ValidatorService._sign_with_key()`
- **Key advancement**: Verify key is advanced until the target slot is in the prepared interval
- **Registry update**: Verify the updated entry is persisted back to the registry
- **Both key types**: Test with both `"attestation_secret_key"` and `"proposal_secret_key"`

#### `ValidatorService._maybe_produce_block()` (async)
- **Proposer match**: When our validator is the proposer, produces and emits a block
- **Not proposer**: No block produced when another validator is scheduled
- **No head state**: Returns early without producing
- **Production failure (AssertionError)**: Logs and skips gracefully

#### `ValidatorService._produce_attestations()` (async)
- **All validators attest**: Each registered validator produces an attestation
- **Block wait logic**: Test the polling loop that waits for the current slot's block
- **Local processing**: Attestation is processed locally before publishing (gossipsub doesn't self-deliver)
- **Duplicate prevention**: Same slot doesn't produce duplicate attestations
- **Slot pruning**: `_attested_slots` is pruned to bound memory

#### `ValidatorService.run()` (async, main loop)
- **Interval routing**: Interval 0 triggers block production, interval ≥ 1 triggers attestation
- **Empty registry**: Loop continues without producing
- **Sleep behavior**: Sleeps until next interval when current is already handled

## Why this matters

- **Block production correctness**: The service decides _when_ and _how_ blocks are signed. Bugs here could produce invalid blocks, double-sign, or miss slots.
- **Key management safety**: The registry loads and manages XMSS secret keys. Incorrect loading (wrong file, wrong index) could sign with the wrong key.
- **OTS key advancement**: `_sign_with_key()` advances one-time signature keys. A bug could exhaust keys prematurely or reuse them.
- **Async timing logic**: The main loop and attestation wait logic involve subtle timing — easy to introduce races or missed intervals.

### Running tests with coverage

```bash
# Registry only
uv run pytest tests/lean_spec/subspecs/validator/test_registry.py -v \
  --cov=src/lean_spec/subspecs/validator/registry --cov-report=term-missing

# Service only
uv run pytest tests/lean_spec/subspecs/validator/test_service.py -v \
  --cov=src/lean_spec/subspecs/validator/service --cov-report=term-missing

# Both
uv run pytest tests/lean_spec/subspecs/validator/ -v \
  --cov=src/lean_spec/subspecs/validator --cov-report=term-missing
```

Target: **≥90% line and branch coverage** for registry, **≥80%** for service (async main loop is harder to cover fully).

### Verifying with the coverage gate

```bash
uvx tox -e coverage-gate
```

### Test file locations

```
tests/lean_spec/subspecs/validator/test_registry.py
tests/lean_spec/subspecs/validator/test_service.py
```

### Testing tips

- **Registry tests** are straightforward — use `tmp_path` for YAML files and mock SSZ key files
- **Service tests** require mocking `SyncService`, `SlotClock`, and the forkchoice `Store`. Use `pytest-asyncio` for async tests
- **`_sign_with_key`** needs real or mocked XMSS keys — check existing test fixtures in `tests/` for patterns
- Registry tests are a natural **good first issue** entry point; service tests are more involved

## Using Claude Code subagents

If you're tackling this issue with Claude Code, use these agents for an efficient workflow:

### 1. `code-tester` agent — Generate the tests
Use this agent to scaffold comprehensive tests. Prompt example:
> Generate unit tests for `src/lean_spec/subspecs/validator/registry.py`. Cover `ValidatorManifestEntry` field validation, `ValidatorManifest.from_yaml_file()`, `load_node_validator_mapping()`, and all `ValidatorRegistry` methods including `from_yaml()` with tmp_path YAML fixtures. Also generate async tests for `src/lean_spec/subspecs/validator/service.py` covering `_sign_block`, `_sign_attestation`, `_maybe_produce_block`, `_produce_attestations`, and the main `run()` loop.

### 2. `doc-writer` agent — Document the test modules
After tests are written, use this agent to add educational documentation. Prompt example:
> Add documentation to the validator test files explaining the testing strategy: why YAML loading edge cases matter, why key advancement must be tested, and what security properties each test group validates.

### Workflow
1. Start with `test_registry.py` (simpler, good first issue material)
2. Run coverage, iterate until ≥90%
3. Move to `test_service.py` (async, needs mocks)
4. Run coverage, iterate until ≥80%
5. Run `doc-writer` to polish documentation
6. Run `uvx tox -e all-checks` to pass all quality checks

### PR DESCRIPTION
**Pull request by dicethedev**
## Summary

Closes #493.

`validator/registry.py` and `validator/service.py` had 0% test coverage
despite being actively used production modules. This PR adds comprehensive
unit and integration tests for both.

### registry.py
- `ValidatorManifestEntry` — field validation including YAML's integer
  coercion of `0x`-prefixed hex strings, which silently breaks pubkey
  parsing without this test
- `ValidatorManifest.from_yaml_file()` — direct YAML loading, previously
  untested even though it's the only entry point for key metadata
- `load_node_validator_mapping()` — normal loading, empty file, single
  node cases
- `ValidatorRegistry` — full lifecycle: `add/get`, `__contains__`,
  `__len__`, `indices()`, `primary_index()`, `from_yaml()` (happy path,
  unknown node, missing manifest entry, missing key file, corrupt key
  file for both attestation and proposal keys), and `from_secret_keys()`
- `ValidatorEntry` — construction and immutability (frozen dataclass)

### service.py
- `_sign_with_key` — key advancement loop (no advance, one advance,
  multiple advances), registry persistence after advancement, and
  isolation between attestation and proposal keys. This is the most
  security-sensitive path: a bug here could exhaust OTS keys or reuse them
- `_sign_block` / `_sign_attestation` — correct field population, correct
  key routing (proposal vs attestation), and `ValueError` for unknown
  validators
- `_maybe_produce_block` — not-proposer early exit, no-head-state early
  exit, and `AssertionError` from the store being swallowed gracefully
- `_produce_attestations` — block-wait polling loop (8 retries, early
  exit when block arrives), local gossip processing before publish,
  counter increments
- `run()` — interval 0 routes to block production, interval ≥ 1 routes
  to attestation, empty registry skips all duties, duplicate slot
  prevention, and `_attested_slots` pruning to bound memory

## Coverage achieved

| File | Before | After |
|---|---|---|
| `validator/registry.py` | 0% | ≥90% |
| `validator/service.py` | 0% | 100% |

## Notes

- `@dataclass(slots=True)` on `ValidatorService` prevents
  `patch.object(instance, method)` — all mocks patch the class and
  include `self` as the first parameter
- `Signature.zero()` is used wherever Pydantic validation rejects
  `MagicMock` for signature fields
- Async tests use `pytest-asyncio` with `mode=auto` already configured
  in `pyproject.toml`
 - Registry tests use `tmp_path` for temporary YAML files and mock key files.
- Service tests mock `SyncService`, `SlotClock`, and `Store`, and use `_zero_sig()` for signature fields.
- Coverage targets: ≥90% for registry, 100% for service (async loops fully exercised).
- Edge cases tested: empty/malformed YAML, missing or corrupted key files, non-proposer block intervals, duplicate attestations, and OTS key exhaustion.
- All quality checks pass: `uvx tox -e all-checks` (`ruff`, `ty`, `codespell`, `mdformat`).
- Tests are safe: no real cryptography is performed.

### CODE REVIEW

File: tests/lean_spec/subspecs/validator/test_registry.py [hunk 1]
```diff
@@ -1,4 +1,22 @@
...
+   We test every edge of that parsing, including YAML's integer coercion.
+
+2. **Key loading and registry lifecycle** — ValidatorRegistry.from_yaml()
+   orchestrates file I/O, manifest lookup, and SSZ key decoding.  We test
+   the happy path plus every failure mode (unknown node, missing manifest
+   entry, missing file, corrupt file) for both attestation and proposal keys.
+
+All YAML files are written to pytest's tmp_path to keep tests hermetic.
+SecretKey.decode_bytes is patched wherever we don't need real key material.
+"""
```
Comment (tcoratger) [inline tests/lean_spec/subspecs/validator/test_registry.py]:
```suggestion
"""Tests for Validator Registry."""
```
